File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int len = nums.size();
        vector<vector<int>> res;

        sort(nums.begin(), nums.end());

        for (int i = 0; i < len - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int a = nums[i];

            int l = i + 1;
            int r = len - 1;

            while (l < r) {
                int b = nums[l];
                int c = nums[r];
                if (a + b + c < 0) {
                    l++;
                } else if (a + b + c > 0) {
                    r--;
                } else {
                    res.push_back({nums[i], nums[l], nums[r]});
                    l++;
                    r--;

                    while (0 < l && l < len && (nums[l] == nums[l - 1])) {
                        l++;
                    }

                    while (0 <= r && r < len - 1 && (nums[r] == nums[r + 1])) {
                        r--;
                    }
                }
            }
        }

        return res;
    }
};

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        int len = nums.length;

        List<List<Integer>> ans = new ArrayList<>();
        for (int i = 0; i < len - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int beg = i + 1, end = len - 1;
            int req = -nums[i];

            while (beg < end) {
                if (nums[beg] + nums[end] == req) {
                    ans.add(List.of(nums[i], nums[beg], nums[end]));
                    // System.out.println("Found: " + List.of(nums[i], nums[beg], nums[end]));
                    beg++;
                    end--;

                    while (beg < end && nums[beg] == nums[beg - 1]) {
                        beg++;
                    }
                    
                    while (beg < end && nums[end] == nums[end + 1]) {
                        end--;
                    }
                } else if (nums[beg] + nums[end] < req) {
                    beg++;
                } else {
                    end--;
                }
            }
        }

        return ans;
    }
}